bookkeeper-server/src/main/java/org/apache/bookkeeper/client/SyncCallbackUtils.java

```diff
@@ -0,0 +1,278 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.bookkeeper.client;
+
+import com.google.common.collect.Iterators;
+import java.util.Enumeration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import static org.apache.bookkeeper.client.LedgerHandle.LOG;
+import org.apache.bookkeeper.client.api.Handle;
+import org.apache.bookkeeper.client.api.ReadHandle;
+
+/**
+ * Utility for callbacks
+ *
+ */
+class SyncCallbackUtils {
+
+    /**
+     * Wait for a result. This is convenience method to implement callbacks
+     *
+     * @param <T>
+     * @param future
+     * @return
+     * @throws InterruptedException
+     * @throws BKException
+     */
+    public static <T> T waitForResult(CompletableFuture<T> future) throws InterruptedException, BKException {
+        try {
+            return future.get();
+        } catch (ExecutionException err) {
+            if (err.getCause() instanceof BKException) {
+                throw (BKException) err.getCause();
+            } else {
+                BKException unexpectedConditionException
+                    = BKException.create(BKException.Code.UnexpectedConditionException);
+                unexpectedConditionException.initCause(err.getCause());
+                throw unexpectedConditionException;
+            }
+
+        }
+    }
+
+    /**
+     * Handle the Response Code and transform it to a BKException
+     *
+     * @param <T>
+     * @param rc
+     * @param result
+     * @param future
+     */
+    public static <T> void finish(int rc, T result, CompletableFuture<T> future) {
+        if (rc != BKException.Code.OK) {
+            future.completeExceptionally(BKException.create(rc).fillInStackTrace());
+        } else {
+            future.complete(result);
+        }
+    }
+
+    static class SyncCreateCallback implements AsyncCallback.CreateCallback {
+
+        private final CompletableFuture future;
+
+        public SyncCreateCallback(CompletableFuture future) {
+            this.future = future;
+        }
+
+        /**
+         * Create callback implementation for synchronous create call.
+         *
+         * @param rc return code
+         * @param lh ledger handle object
+         * @param ctx optional control object
+         */
+        @Override
+        @SuppressWarnings(value = "unchecked")
+        public void createComplete(int rc, LedgerHandle lh, Object ctx) {
+            SyncCallbackUtils.finish(rc, lh, future);
+        }
+
+    }
+
+    static class SyncOpenCallback<T> implements AsyncCallback.OpenCallback {
+
+        private final CompletableFuture future;
```

Comment:
doesn't CompletableFuture<? super LedgerHandle> work? I am just curious -- I was suggesting adding type to ensure the right completable future is passed in.